src/EditorFeatures/CSharp/CodeCleanup/CSharpCodeCleanupService.cs

```diff
@@ -0,0 +1,181 @@
+                                          IDEDiagnosticIds.UseCollectionInitializerDiagnosticId)
+                }
+            };
+
+            return dictionary.ToImmutableDictionary();
+        }
+
+        public async Task<Document> CleanupDocument(Document document, CancellationToken cancellationToken)
+        {
+            var docOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            document = await ApplyCodeFixes(document, docOptions, cancellationToken).ConfigureAwait(false);
+            // do the remove usings after code fix, as code fix might remove some code which can results in unused usings.
+            document = await RemoveSortUsings(document, docOptions, cancellationToken).ConfigureAwait(false);
+
+            return document;
+        }
+
+        private async Task<Document> RemoveSortUsings(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        {
+            // remove usings
+            if (docOptions.GetOption(FeatureOnOffOptions.RemoveUnusedUsings))
+            {
+                var removeUsingsService = document.GetLanguageService<IRemoveUnnecessaryImportsService>();
+                if (removeUsingsService != null)
+                {
+                    document = await removeUsingsService.RemoveUnnecessaryImportsAsync(document, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            // sort usings
+            if (docOptions.GetOption(FeatureOnOffOptions.SortUsings))
+            {
+                document = await OrganizeImportsService.OrganizeImportsAsync(document, cancellationToken).ConfigureAwait(false);
+            }
+
+            return document;
+        }
+
+        private async Task<Document> ApplyCodeFixes(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        {
+            var fixAllService = document.Project.Solution.Workspace.Services.GetService<IFixAllGetFixesService>();
+
+            var dummy = new ProgressTracker();
+            foreach (var diagnosticId in GetEnabledDiagnosticIds(docOptions))
+            {
+                var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var textSpan = new TextSpan(0, syntaxTree.Length);
+
+                var fixCollection = await _codeFixService.GetFixesAsync(document, textSpan, diagnosticId, cancellationToken).ConfigureAwait(false);
+                if (fixCollection == null)
+                {
+                    continue;
+                }
+
+                var fixAll = fixCollection.FixAllState;
+                var solution = await fixAllService.GetFixAllChangedSolutionAsync(
+                    fixAll.CreateFixAllContext(dummy, cancellationToken)).ConfigureAwait(false);
+
+                document = solution.GetDocument(document.Id);
```

Comment:
this approach seems pretty expensive overall.  If we have N operations to perform, we end up forking the solution N times, and having to do all the semantic work over and over again each time.  It feels like it might be more ideal to just compute all N operations against the *original* document, then attempt to merge in teh results instead of forking N times.

Reply:
Tracked in https://github.com/dotnet/roslyn/issues/27346